added creation of the Balance Audit table,

Ramil23-bit opened the pull request:
created an annotation to start saving the account record to the table

Ramil23-bit committed:
added creation of the Balance Audit table, created an annotation to start saving the account record to the table

20cognosce commented on src/main/java/faang/school/accountservice/aspect/BalanceAuditAbstractAspect.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class BalanceAuditAbstractAspect extends BalanceAuditAbstract {
```
почему назвал BalanceAuditAbstractAspect, если он не абстрактный

и зачем ты вообще выделял абстрактный класс

и где это аспект? я вижу только, что это бин

20cognosce commented on src/main/java/faang/school/accountservice/aspect/BalanceAuditAbstract.java:
```diff
@@ -0,0 +1,28 @@
+import java.time.LocalDateTime;
+import java.util.UUID;
+@RequiredArgsConstructor
+public abstract class BalanceAuditAbstract {
```
Abstract что? это сервис ведь?

20cognosce commented on src/main/java/faang/school/accountservice/aspect/BalanceAuditCreateAspect.java:
```diff
@@ -0,0 +1,34 @@
+        }
+        Object result = joinPoint.proceed();
+
+        UUID accountId = account.getId();
```
откуда взялась переменная account, это компилируется вообще?

20cognosce commented on src/main/java/faang/school/accountservice/aspect/BalanceAuditCreateAspect.java:
```diff
@@ -0,0 +1,34 @@
+        if (args.length == 0 || !(args[0] instanceof Account account)) {
+            throw new IllegalStateException("Method annotated with @BalanceAuditCreateAnnotation must have an Account parameter");
+        }
+        Object result = joinPoint.proceed();
```
при изменении баланса - сначала берешь снимок до выполнения метода, затем вызываешь proceed, и затем берешь снимок после

если есть время делать для изменения баланса

20cognosce commented on src/main/java/faang/school/accountservice/controller/BalanceAuditController.java:
```diff
@@ -0,0 +1,13 @@
+@RequestMapping("/api/v1/balance_audit")
+@RequiredArgsConstructor
+public class BalanceAuditController {
+
```
Почему пустой контроллер

20cognosce commented on src/main/java/faang/school/accountservice/controller/BalanceAuditController.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v1/balance_audit")
```
нижнего подчеркивания не должно быть в API
balance-audit

20cognosce commented on src/main/resources/db/changelog/changeset/V003__balance-audit-table-init.sql:
```diff
@@ -0,0 +1,9 @@
+    authorized_amount NUMERIC(30, 10) NOT NULL,
+    actual_amount NUMERIC(30, 10) NOT NULL,
+    balance_change_id bigint,
+    created_at TIMESTAMP NOT NULL DEFAULT current_timestamp
```
Форматирование вертикальное

Ramil23-bit replied:
@20cognosce Если честно не очень понимаю что нужно сделать. вроде все отформатировано

20cognosce replied:
> @20cognosce Если честно не очень понимаю что нужно сделать. вроде все отформатировано

<img width="1174" height="558" alt="image" src="https://github.com/user-attachments/assets/815824b3-7c88-450f-b882-5e216884e1cc" />

20cognosce commented on src/main/resources/db/changelog/changeset/V003__balance-audit-table-init.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE balance_audit (
+    id bigint PRIMARY KEY GENERATED ALWAYS AS IDENTITY UNIQUE,
+    account_number varchar(20) UNIQUE NOT NULL,
```
где внешний ключ на аккаунт, который ты аудировал

20cognosce commented on src/main/java/faang/school/accountservice/service/balance_audit/BalanceAuditService.java:
```diff
@@ -0,0 +1,32 @@
+    public Balance searchBalance(UUID accountId) {
+        Optional<Balance> balanceOptional = balanceRepository.findByAccountId(accountId);
+        if (balanceOptional.isEmpty()) {
+            throw new DataValidationException("This account does not exist");
```
Это не ошибка валидации, это NotFoundException

20cognosce commented on src/main/java/faang/school/accountservice/entity/BalanceAudit.java:
```diff
@@ -0,0 +1,66 @@
+
+  - include:
+      file: db/changelog/changeset/V003__balance-audit-table-init.sql
+ */
```
что это

20cognosce commented on src/main/resources/db/changelog/changeset/V003__balance-audit-table-init.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE balance_audit (
```
И где самое главное - КТО СОВЕРШИЛ ЭТО ДЕЙСТВИЕ?

Ramil23-bit replied:
@20cognosce в БД будет записываться номер аккаунта, этого будет мало?? или нужно прям ФИО?)))

20cognosce replied:
> @20cognosce в БД будет записываться номер аккаунта, этого будет мало?? или нужно прям ФИО?)))

Желательно связь с user_id

20cognosce commented on src/main/java/faang/school/accountservice/entity/BalanceAudit.java:
```diff
@@ -0,0 +1,66 @@
+@AllArgsConstructor
+@Entity
+@Table(name = "balance_audit")
+public class BalanceAudit {
```
Ты не аудируешь инициатора действия

20cognosce reviewed (changes requested):
Сырой, полурабочий код

Нужно доработать

Ramil23-bit committed:
fixed code.

"И где самое главное - КТО СОВЕРШИЛ ЭТО ДЕЙСТВИЕ?"
This comment didn't quite understand what needs to be done?

20cognosce commented on src/main/resources/db/changelog/changeset/V003__balance-audit-table-init.sql:
```diff
@@ -0,0 +1,10 @@
+    balance_version BIGINT NOT NULL,
+    authorized_amount NUMERIC(30, 10) NOT NULL,
+    actual_amount NUMERIC(30, 10) NOT NULL,
+    balance_change_id BIGINT,
```
ну и где ты заполняешь в коде это поле

20cognosce replied:
на что оно вообще указывает

Ramil23-bit replied:
@20cognosce я пока это поле не заполняю вообще, потому что оно заполняется после того как ребята дмс сделают, мне его сейчас даже заполнить не чем

20cognosce commented on src/main/java/faang/school/accountservice/service/balance_audit/BalanceAuditService.java:
```diff
@@ -0,0 +1,46 @@
+        balanceAudit.setBalanceVersion(balance.getVersion().longValue());
+
+        return balanceAuditRepository.save(balanceAudit);
+    }
```
user_id инициатора действия заполни

20cognosce commented:
Конфликт

Ramil23-bit committed:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-78130_balance_audit

# Conflicts:
#	src/main/resources/db/changelog/db.changelog-master.yaml

Ramil23-bit committed:
fixed code.

Added user_id to table balance audit

20cognosce commented on src/main/java/faang/school/accountservice/service/balance_audit/BalanceAuditService.java:
```diff
@@ -0,0 +1,47 @@
+        return balanceOptional.get();
+    }
+
+    public BalanceAudit saveBalanceAudit(UUID accountId) {
```
в транзакции надо это делать

20cognosce commented on src/main/java/faang/school/accountservice/service/balance_audit/BalanceAuditService.java:
```diff
@@ -0,0 +1,47 @@
+        balanceAudit.setActualAmount(balance.getActualAmount());
+        balanceAudit.setCreatedAt(LocalDateTime.now());
+        balanceAudit.setBalanceVersion(balance.getVersion().longValue());
+        balanceAudit.setUserId(balance.getAccount().getUserId());
```
из контекста надо доставать, а не из аккаунта
